### ISSUE TITLE
Thyra Tpetra Adapters: TpetraVectorSpace to override smallVecSpcFcty()

### ISSUE THREAD

**Issue by bartgol** [trilinos/Trilinos#3237]
@trilinos/thyra

## Expectations

The method `smallVecSpcFcty()` should return a factory that creates a vector space of TpetraVectorSpace type.

## Current Behavior

The class `TpetraVectorSpace` does not override this method, relying on the implementation of the mother class `SpmdVectorSpaceDefaultBase`.

## Motivation and Context

The current implementation is ok as long as one only works at the level of Thyra interfaces. The implementation in `SpmdVectorSpaceDefaultBase` returns a default factory which creates a default locally replicated vector space, which is fine at the thyra level. However, things get tricky when diving in the adapters world, inside an external application. Most likely, deep down the stack, the customer application casts most of the thyra objects to a concrete implementation. In case of Tpetra adapters, the app will cast, say, a `Thyra::VectorBase` to a `Thyra::TpetraVector`, then extract the Tpetra vector and do stuff with it. This would be a problem in case the vector was generated by the vector space generated by `Thyra::TpetraVectorSpace::smallVecSpcFcty()`, which is _not_ a `Thyra::TpetraVectorSpace`.

## Possible Solution

This is what I would do:
- create a `TpetraVectorSpaceFactory` which creates a locally replicated `Tpetra::Map`, which then wraps in a `TpetraVectorSpace`.
- in `TpetraVectorSpace`, override the `smallVecSpcFcty()` method, to return a `TpetraVectorSpaceFactory`.

## Additional Information

Note: this _could_ (in principle) break existing code, since the `TpetraVectorSpaceFactory` would create a `TpetraVectorSpace`, which would then create objects of type `TpetraVector`, which inherit from `SpmdVectorDefaultBase`. The current `smallVecSpcFcty()` in `TpetraVectorSpace` (which is implemented in `SpmdVectorDefaultBase`) creates spaces of `DefaultSpmdVectorSpace` type, which then create `DefaultSpmdVector`, which is a sibling (but not parent) to `TpetraVector` (as they both inherit from `DefaultSpmdVector`).

This code break would not be related to the proposed modification, per se. It is a consequence of a (imho) current design problem: `SpmdVectorSpaceDefaultBase` creates a `DefaultSpmdVectorSpaceFactory` which then creates vector spaces that are not of type `SpmdVectorSpaceDefaultBase`, but are instead of derived type `DefaultSpmdVectorSpace`. In my opinion, it would be more correct if the vector space factory returned by `SpmdVectorSpaceDefaultBase` was of type, say, `SpmdVectorSpaceDefaultBaseFactory` (this doesn't exist, I'm making the name up), which would create vector spaces of the abstract type `SpmdVectorSpaceDefaultBase`, rather than of the derived type `DefaultSpmdVectorSpace`. The latter, by the way, is more than welcome to override this.

In short: I think a vector space of type X should create a vs factory that creates vector spaces of type X. This ensures that the small spaces created by the factory created by vs X are fully compatible with the domain spaces of multivectors created by vs X.

Sorry for the super long issue. Thanks for enduring the long read!

**Comment by bartgol** [comment]
Note: right now, in order to create a small tpetra vector space (locally replicated) from a tpetra vector space, I need to do:

`auto small_vs = Thyra::createMembers(tpetra_vs, numMembers)->domain()`

which is a waste (creates multivectors for nothing), rather than the much cleaner

`auto small_vs = tpetra_vs->smallVecSpcFcty()->createVecSpc(numMembers)`

**Comment by bartgol** [comment]
Note 2: I would be willing to put up a PR for this, but the issue with backward compatibility holds me back. I think this requires some design thinking, and I think I need the opinion of thyra designers (@bartlettroscoe , I'm pinging you cause you're the thyra POC).

**Comment by jhux2** [comment]
@trilinos/thyra

**Comment by mhoemmen** [comment]
@trilinos/tpetra

**Comment by bartlettroscoe** [comment]
What is the underlying use case that is bringing this up?  Instead of having APP codes getting Tpetra objects for everything, perhaps some FACADE could be used to return Kokkos::View or or Teuchos::ArrayRCP views of the data?

I am worried about the performance hit making this change would cause.

**Comment by bartgol** [comment]
@bartlettroscoe Thanks for the feedback. And sorry if I'm pinging you a lot lately; we're pushing a lot for extensive use of Thyra in Albany, so I'm getting deeper and deeper in it (heads up: I may have more doubts in the next days/weeks).

I am indeed using a facade that extracts (multi)vector's data, returning ArrayRCP (or views). The reason the current issue showed up is that at some point Albany builds a locally replicated vector. I am not 100% sure this is correct, but that made me find out I had no way to create a locally replicated Thyra::TpetraVector using just thyra abstraction (i.e., no explicit knowledge of underlying linear algebra) unless I a) created a multivector (with a proper domain vs dim), and b) extracted its domain space.

As for the multivector creation, I _think_ the tpetra adapters already build local maps every time they create multivectors (be it from the `creteMembers` function or from the constructor-initialize methods of `TpetraMultiVector`), so I _think_ this would not change the overhead costs of multivectors constructions (at least for the Tpetra adapters case, not sure about the other VS types that derive from `SpmdVectorSpaceDefaultBase`).

I guess my main concern about the current design is that, given a vector space vs of type X (where X inherits from `SpmdVectorSpaceDefaultBase`), doing `vs->smallVecSpcFcty()->createVecSpc()` yields a vector space that is no longer of type X (nor of any parent of X). I _think_ this could lead to some inefficiencies (or even bugs) down the road.

E.g., (not that it already happens anywhere (that I know of)) what if someone has a TpetraMultiVector x , creates a locally replicated multivector y using the small vector space factory (hence doesn't get back a TpetraMultiVector), then uses the apply method of x on y (this is legal, since the MV is a LinearOp). The Tpetra implementation will try first to cast to TpetraMultiVector, and since it will fail, it will fall back to RTOp implementation (which is on host). Point is, not having small vector spaces be of the same concrete implementation of the original VS _could_ cause loss of efficiency down the road (with a good adapter impl, we fall back on less efficient generic implementations, with a bad (or buggy) impl we may get an error).

That said, I am not in desperate need of this feature. As I said, I can get around it by creating a multivector, then grabbing its domain space. But I thought this deserved some sort of discussion, to see if it is the behavior we want or if it needs fixing.

**Comment by bartlettroscoe** [comment]
@bartgol, okay, we should look into this a little deeper.  But I agree that it would be more consistent and less confusing if every Thyra::VectorSpaceBase object and every Thyr::MultiVectorBase created from a chain of calls to a Thyra::TpetraVectorSpace object be Thyra::TpetraVectorSpace and Thyra::TpetraMultiVector objects.  But we need some type of performance test to see how this impacts things.

As for the comment:

> I may have more doubts in the next days/weeks

we should talk through the major concerns there.  It may just be that Thyra is not really the right interface for what you are trying to achieve in these Albany refactorings.  I fear the correct interface, an APP-LAL interface, does not really exist in Trilinos (unless you focus on just Epetra and Tpetra and then you could use Xpetra).

**Comment by bartgol** [comment]
> we should talk through the major concerns there. It may just be that Thyra is not really the right interface for what you are trying to achieve in these Albany refactorings. I fear the correct interface, an APP-LAL interface, does not really exist in Trilinos (unless you focus on just Epetra and Tpetra and then you could use Xpetra).

I guess we can take that discussion offline.
